### PR TITLE
Batch

### PR DESCRIPTION
**Pull request by vishalbollu**
---

checklist:

- [x] run `make test` and `make lint`
- [x] test manually (i.e. build/push all images, restart operator, and re-deploy APIs)
- [x] update examples
- [x] update docs and add any new files to `summary.md` (view in [gitbook](https://docs.cortex.dev/v/master) after merging)
- [ ] cherry-pick into release branches if applicable
- [ ] alert the dev team if the dev environment changed

### CODE REVIEW

File: pkg/operator/resources/batchapi/cleanup_cron.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+		err := deleteJobRuntimeResources(
+			spec.JobKey{APIName: k8sJobMap[jobID].Labels["apiName"], ID: k8sJobMap[jobID].Labels["jobID"]},
+		)
+		if err != nil {
+			telemetry.Error(err)
+			errors.PrintError(err)
+		}
+	}
+
+	for jobID := range strset.Difference(jobIDSetQueueURL, jobIDSetK8sJob, inProgressJobIDSet) {
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/cleanup_cron.go:100]: Handle case where queue exists but enquing state file has not been written yet

File: pkg/lib/aws/sqs.go [hunk 2]
```diff
@@ -22,15 +22,32 @@ import (
...
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to get queue attributes", queueURL)
+	}
+
+	return aws.StringValueMap(output.Attributes), nil
+}
+
 func (c *Client) ListQueuesByQueueNamePrefix(queueNamePrefix string) ([]string, error) {
-	output, err := c.SQS().ListQueues(&sqs.ListQueuesInput{
+	queueURLs := []string{}
```
Comment (deliahu) [inline pkg/lib/aws/sqs.go:38]: nit: `var queueURLs []string`

File: pkg/operator/resources/resources.go [hunk 3]
```diff
@@ -39,19 +39,22 @@ import (
 	kcore "k8s.io/api/core/v1"
 )
```
Comment (deliahu) [inline pkg/operator/resources/resources.go:41]: nit: Add a comment here that says that this function returns an error if the resource does not exist

File: pkg/operator/resources/batchapi/api.go [hunk 4]
```diff
@@ -161,12 +161,13 @@ func deleteS3Resources(apiName string) error {
 	)
 }
 
+// Returns all batch apis, for each API returning the most recently submitted job and all running APIs
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/api.go:164]: Should "running APIs" be "running jobs"?

File: pkg/operator/resources/batchapi/job_state.go [hunk 5]
```diff
@@ -131,11 +135,12 @@ func getJobStateFromFiles(jobKey spec.JobKey, lastUpdatedFileMap map[string]time
 
 func getMostRecentlySubmittedJobStates(apiName string, count int) ([]*JobState, error) {
 	// a single job state may include 5 files on average, overshoot the number of files needed
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/job_state.go:137]: should this comment be moved to the const above?

File: pkg/operator/resources/batchapi/job_state.go [hunk 6]
```diff
@@ -335,15 +337,16 @@ func getJobStatusFromJobState(jobState *JobState, k8sJob *kbatch.Job, pods []kco
 				setUnexpectedErrorStatus(jobKey)
 				deleteJobRuntimeResources(jobKey)
 				jobStatus.Status = status.JobUnexpectedError
+				return &jobStatus, nil
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/job_state.go:340]: should this return an error?

File: pkg/operator/resources/batchapi/errors.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+	ErrMessageExceedsMaxSize      = "batchapi.message_exceeds_max_size"
+	ErrConflictingFields          = "batchapi.conflicting_fields"
+	ErrBatchItemSizeExceedsLimit  = "batchapi.item_size_exceeds_limit"
+	ErrSpecifyExactlyOneKey       = "batchapi.specify_exactly_one_key"
+)
+
+func ErrorBatchAPINotDeployed(apiName string) error {
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrBatchAPINotDeployed,
+		Message: fmt.Sprintf("BatchAPI api named '%s' is not deployed", apiName),
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/errors.go:44]: consider "BatchAPI api" -> "batch api" (we can discuss)

File: pkg/operator/resources/batchapi/errors.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+	})
+}
+
+func ErrorJobNotFound(jobKey spec.JobKey) error {
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrJobNotFound,
+		Message: fmt.Sprintf("unable to find job %s", jobKey.UserString()),
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/errors.go:51]: consider "job" -> "batch job"

File: pkg/operator/resources/batchapi/errors.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+	})
+}
+
+func ErrorJobIsNotInProgress() error {
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrJobIsNotInProgress,
+		Message: "cannot stop job because it is not in progress",
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/errors.go:58]: consider "job" -> "batch job"

File: pkg/operator/resources/batchapi/errors.go [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrJobHasAlreadyBeenStopped,
+		Message: "job has already been stopped",
+	})
+}
+
+func ErrorNoS3FilesFound() error {
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrNoS3FilesFound,
+		Message: "no s3 files found based on search criteria",
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/errors.go:72]: consider "no s3 files match search criteria"

File: pkg/operator/resources/batchapi/errors.go [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrNoDataFoundInJobSubmission,
+		Message: "unable to enqueue batches because no data was found",
+	})
+}
+
+func ErrorFailedToEnqueueMessages(message string) error {
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrFailedToEnqueueMessages,
+		Message: fmt.Sprintf(message),
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/errors.go:86]: `fmt.Sprintf` is unnecessary

File: pkg/operator/resources/batchapi/errors.go [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+		Kind:    ErrMessageExceedsMaxSize,
+		Message: fmt.Sprintf("cannot enqueue message because its size of %d bytes exceeds the %d bytes limit; use a smaller batch size or reduce the size of each of item in the batch", messageSize, messageLimit),
+	})
+}
+
+func ErrorConflictingFields(key string, keys ...string) error {
+	allKeys := append(keys, key)
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrConflictingFields,
+		Message: fmt.Sprintf("please specify either the %s field (both not more than one at the same time)", s.StrsOr(allKeys)),
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/errors.go:101]: "both" -> "but"

File: pkg/operator/resources/batchapi/errors.go [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+
+func ErrorMessageExceedsMaxSize(messageSize int, messageLimit int) error {
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrMessageExceedsMaxSize,
+		Message: fmt.Sprintf("cannot enqueue message because its size of %d bytes exceeds the %d bytes limit; use a smaller batch size or reduce the size of each of item in the batch", messageSize, messageLimit),
+	})
+}
+
+func ErrorConflictingFields(key string, keys ...string) error {
+	allKeys := append(keys, key)
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/errors.go:98]: nit: I think it's good to preserve the order:

```go
allKeys := append([]string{key}, keys...)
```

File: pkg/operator/resources/batchapi/errors.go [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrConflictingFields,
+		Message: fmt.Sprintf("please specify either the %s field (both not more than one at the same time)", s.StrsOr(allKeys)),
+	})
+}
+
+func ErrorItemSizeExceedsLimit(index int, size int, limit int) error {
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrBatchItemSizeExceedsLimit,
+		Message: fmt.Sprintf("item %d has size %d bytes which exceeds the limit %d", index, size, limit),
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/errors.go:108]: nit: "which exceeds the limit (%d bytes)"

File: pkg/operator/resources/batchapi/errors.go [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+	})
+}
+
+func ErrorSpecifyExactlyOneKey(key string, keys ...string) error {
+	allKeys := append(keys, key)
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/errors.go:113]: nit: I think it's good to preserve the order:

```go
allKeys := append([]string{key}, keys...)
```

File: pkg/operator/resources/batchapi/errors.go [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+	return errors.WithStack(&errors.Error{
+		Kind:    ErrSpecifyExactlyOneKey,
+		Message: fmt.Sprintf("specify exactly one of the following keys %s", s.StrsOr(allKeys)), // TODO add job specification documentation
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/errors.go:116]: nit: either "specify exactly one of the following keys: %s" or "specify exactly one of %s"

File: pkg/operator/resources/batchapi/k8s_specs.go [hunk 17]
```diff
@@ -0,0 +1,224 @@
...
+	}
+}
+
+func pythonPredictorJobSpec(api *spec.API, job *spec.Job) (*kbatch.Job, error) {
+	containers, volumes := operator.PythonPredictorContainers(api)
+	for i, container := range containers {
+		if container.Name == operator.APIContainerName {
+			containers[i].Env = append(container.Env, kcore.EnvVar{
+				Name:  "CORTEX_JOB_SPEC",
+				Value: "s3://" + config.Cluster.Bucket + "/" + job.FileSpecKey(),
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/k8s_specs.go:55]: consider renaming `FileSpecKey()` to `SpecKey()` or `SpecFile()` or `SpecFileKey()`

File: pkg/operator/resources/batchapi/k8s_specs.go [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+			})
+		}
+	}
+
+	return k8s.Job(&k8s.JobSpec{
+		Name:        job.JobKey.K8sName(),
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/k8s_specs.go:61]: The name `JobKey` could be misleading (due to S3 Keys), but I can't think of a better proposal right now. We can discuss

File: pkg/operator/resources/batchapi/k8s_specs.go [hunk 19]
```diff
@@ -0,0 +1,224 @@
...
+				},
+				Tolerations:        operator.Tolerations,
+				Volumes:            operator.DefaultVolumes,
+				ServiceAccountName: "default",
+			},
+		},
+	}), nil
+}
+
+func virtualServiceSpec(api *spec.API) *istioclientnetworking.VirtualService {
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/k8s_specs.go:191]: (reminder to review this one together)

File: pkg/operator/resources/batchapi/k8s_specs.go [hunk 20]
```diff
@@ -0,0 +1,224 @@
...
+		_, err := config.K8s.CreateVirtualService(newVirtualService)
+		return err
+	}
+
+	_, err := config.K8s.UpdateVirtualService(prevVirtualService, newVirtualService)
+	return err
+}
+
+func getVirtualService(apiName string) (*istioclientnetworking.VirtualService, error) {
+	virtualService, err := config.K8s.GetVirtualService(operator.K8sName(apiName))
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/k8s_specs.go:222]: this can be made into one line

File: pkg/operator/resources/batchapi/validations.go [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+	}
+
+	if submission.ItemList != nil {
+		if len(submission.ItemList.Items) == 0 {
+			return errors.Wrap(cr.ErrorTooFewElements(1), schema.ItemsKey)
+		}
+
+		for i, batch := range submission.ItemList.Items {
+			if len(batch) > _messageSizeLimit {
+				return ErrorItemSizeExceedsLimit(i, len(batch), _messageSizeLimit)
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/validations.go:56]: Do we check later that the total size of one batch is not too big?

File: pkg/operator/resources/batchapi/validations.go [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+	"github.com/aws/aws-sdk-go/service/s3"
+	awslib "github.com/cortexlabs/cortex/pkg/lib/aws"
+	cr "github.com/cortexlabs/cortex/pkg/lib/configreader"
+	"github.com/cortexlabs/cortex/pkg/lib/errors"
+	"github.com/cortexlabs/cortex/pkg/lib/pointer"
+	"github.com/cortexlabs/cortex/pkg/operator/schema"
+	"github.com/gobwas/glob"
+)
+
+func jobSubmissionSchemaValidation(submission *schema.JobSubmission) error {
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/validations.go:29]: nit: consider renaming to `validateJobSubmissionSchema()`

File: pkg/operator/resources/batchapi/validations.go [hunk 23]
```diff
@@ -0,0 +1,179 @@
...
+			filesFound++
+			return false, nil
+		})
+
+		if err != nil {
+			return errors.Wrap(err, s3Path)
+		}
+	}
+
+	if filesFound == 0 {
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/validations.go:150]: Do you intend to check that each s3Path has files or that any do? It seems to be checking the later. If that's what you intend, we can short-circuit the loop on s3Paths. If that's not what you indented, I think the file checking should be moved to within that loop

File: pkg/operator/resources/batchapi/validations.go [hunk 24]
```diff
@@ -0,0 +1,179 @@
...
+		return ErrorNoS3FilesFound()
+	}
+
+	return nil
+}
+
+func listFilesDryRun(s3Lister *schema.S3Lister) ([]string, error) {
+	s3Files := []string{}
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/validations.go:158]: nit: `var s3Files []string`

File: pkg/workloads/cortex/serve/batch.py [hunk 25]
```diff
@@ -0,0 +1,343 @@
...
+    if "headers" in local_cache["predict_fn_args"]:
+        args["headers"] = None
+    if "query_params" in local_cache["predict_fn_args"]:
+        args["query_params"] = None
+    if "batch_id" in local_cache["predict_fn_args"]:
+        args["batch_id"] = batch_id
+    return args
+
+
+def renew_message_visibility(queue_url, receipt_handle, initial_offset, interval, *args, **kwargs):
```
Comment (deliahu) [inline pkg/workloads/cortex/serve/batch.py:115]: Double check that this runs for the first time after 60 seconds

File: pkg/types/metrics/job_metrics.go [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+limitations under the License.
+*/
+
+package metrics
+
+type JobMetrics struct {
+	Succeeded           int      `json:"succeeded"`
+	Failed              int      `json:"failed"`
+	AverageTimePerBatch *float64 `json:"average_time_per_batch"`
+	TotalCompleted      int      `json:"total_completed"`
```
[the remark this answers is not in the document]
Comment (vishalbollu) [reply pkg/types/metrics/job_metrics.go:23]: removed

File: pkg/workloads/cortex/serve/batch.py (hunk 25 above)
Comment (vishalbollu) [reply pkg/workloads/cortex/serve/batch.py:115]: Removed

File: pkg/operator/resources/batchapi/job_state.go [hunk 27]
```diff
@@ -298,55 +322,85 @@ func setUnexpectedErrorStatus(jobKey spec.JobKey) error {
...
+	jobKey := initialJobState.JobKey
 
 	jobSpec, err := downloadJobSpec(jobKey)
 	if err != nil {
 		return nil, err
 	}
 
-	statusCode := jobState.Status
+	latestJobState := initialJobState
+
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/job_state.go:334]: add comment for what situation this helps with

File: pkg/operator/resources/batchapi/logs.go [hunk 28]
```diff
@@ -0,0 +1,293 @@
...
+	logGroupName := logGroupNameForJob(jobKey)
+	eventCache := newEventCache(_maxCacheSize)
+	lastLogStreamRefresh := time.Time{}
+	logStreamNames := strset.New()
+	lastLogTime := time.Now()
+	didShowFetchingMessage := false
+	didFetchLogs := false
+	var jobSpec *spec.Job
+
+	jobSpec, err := downloadJobSpec(jobKey)
```
Comment (deliahu) [inline pkg/operator/resources/batchapi/logs.go:163]: Can this be avoided if we pass in the `jobStatus` (which has the start time)?

File: pkg/operator/resources/batchapi/logs.go [hunk 29]
```diff
@@ -0,0 +1,293 @@
...
+	logGroupName := logGroupNameForJob(jobStatus.JobKey)
+
+	logStreamNames, err := getLogStreams(jobStatus.JobKey)
+	if err != nil {
+		telemetry.Error(err)
+		writeAndCloseSocket(socket, "error encountered while searching for log streams: "+errors.Message(err))
+		return
+	}
+
+	endTime := jobStatus.EndTime.Add(time.Minute * 10)
```
[the remark this answers is not in the document]
Comment (vishalbollu) [reply pkg/operator/resources/batchapi/logs.go:95]: Yes that is correct.

[the remark this answers is not in the document]
Comment (vishalbollu) [reply pkg/operator/resources/batchapi/logs.go:95]: jobStatus.EndTime is a pointer so in theory, it should never be in a zero state. Having said that, but I think `time.Now()` should work. Will update.

**Pull request merged**